csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,319 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <abstract_tensor.h>
+#include <abstract_tensor_schedule.h>
+#include <ir/internal_base_nodes.h>
+#include <iter_visitor.h>
+#include <val_graph.h>
+
+namespace nvfuser {
+
+namespace {
+
+class AbstractTensorSchedule {
+ public:
+  static void apply(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph) {
+    AbstractTensorSchedule ats(abstract, concrete, graph);
+    ats.run();
+  }
+
+ private:
+  AbstractTensorSchedule(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph)
+      : abstract_(abstract), concrete_(concrete), graph_(graph) {}
+
+  void run() {
+    findNearestProducers();
+
+    // Now, for each ValGroup in abstract, find the closest producer ValGroups
+    // with entries in concrete_ids_ and replay the path from them. If none
+    // exists, then do not include this dimension in the output
+    std::vector<IterDomain*> loop_domain;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    concrete_->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    NVF_ERROR(
+        abs_id.is<ValGroupAndItsGraph>(),
+        "AbstractId must be a ValGroupAndItsGraph");
+    const ValGroupAndItsGraph& vgg = abs_id.as<ValGroupAndItsGraph>();
+    if (graph_ == nullptr) {
+      graph_ = vgg.graph;
+    } else {
+      NVF_ERROR(graph_ == vgg.graph);
+    }
+    return vgg.group;
+  }
+
+  //! Work backward from each loop IterDomain in concrete. When we find an
+  //! IterDomain* with a ValGroup in scheduled_val_groups, we map to it
+  //! (including siblings) in concrete_ids_. Assert that we do find a
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  //!
+  //! For example, suppose we have the following:
+  //!
+  //!   concrete_:
+  //!     root: iS0{i0} iS1{i1}
+  //!     loop: iS3{ceilDiv(i0 * i1, 16)} iS4{16}
+  //!     transforms:
+  //!       iS2 = merge(iS0, iS1)
+  //!       iS3, iS4 = split(iS2, 16)
+  //!   graph:
+  //!     ValGroup 0: 0 8
+  //!     ValGroup 1: 1 9
+  //!     ValGroup 2: 2 10
+  //!     ValGroup 3: 3
+  //!     ValGroup 4: 4
+  //!     ValGroup 5: 11
+  //!     ValGroup 6: 12
+  //!     ExprGroup 0: iS11{ceilDiv(i0 * i1, 32)}, iS12{32} = split(iS10, 32)
+  //!   abstract_:
+  //!     {{ValGroup 5, graph}, {ValGroup 6, graph}}
+  //!
+  //! Then findNearestProducers() work backward from the loop domain of
+  //! concrete_ (ValGroups 3 and 4) and will identify ValGroup 2 as a producer
```

Comment:
This comment is confusing to me. The code below traverses backward from the abstract domains, not the loop domains of the concrete tensor. What am I missing?

Reply:
We first traverse `abstract_tensor_` backward to get a collection of all the `ValGroup`s that appear in the history of `abstract_tensor_`. Then we traverse `tv_` to find the latest `IterDomain`s that have a `ValGroup` that was found in the first step: https://github.com/NVIDIA/Fuser/blob/88ababfe91e5ad0d4b28946ae490c43b351f286b/csrc/abstract_tensor_schedule.cpp#L127-L128.